addons/resource/models/resource.py

```diff
@@ -987,6 +988,82 @@ def _get_unavailable_intervals(self, start, end):
         return resource_mapping
 
+    def _leaves_interval_gantt_view(self, start_dt, end_dt, leaves, attendances):
+        self.ensure_one()
+
+        result = []
+        tz = timezone(self.user_id.tz or self.calendar_id.tz)
+        start = start_dt.astimezone(tz)
+        end = end_dt.astimezone(tz)
+
+        for leave in leaves:
+            if leave.resource_id.id not in [False, self.id]:
+                continue
+
+            dt0 = string_to_datetime(leave.date_from).astimezone(tz)
+            dt1 = string_to_datetime(leave.date_to).astimezone(tz)
+            result.append((max(start, dt0), min(end, dt1), leave))
+
+        return attendances - Intervals(result)
+
+    def _get_unavailable_intervals_gantt(self, start_datetime, end_datetime):
+
+        leaves_mapping = defaultdict(dict)
+
+        calendars_set = set()
+        for res in self:
+            calendars_set.add(res.calendar_id)
+
+        start_dt = utc.localize(start_datetime) if not start_datetime.tzinfo else start_datetime
+        end_dt = utc.localize(end_datetime) if not end_datetime.tzinfo else end_datetime
+        # both search are made in batch. probably a good idea to mapped the personnal leaves per resource_id in order to not iterate over all leaves every time.
+        personal_leaves_intervals = self._personal_leaves_intervals(start_dt, end_dt)
+        global_leave_per_calendar = self._global_leaves_per_calendar(start_dt, end_dt, calendars_set)
+        dict_attendance_per_calendar = {}
+        for calendar in calendars_set:
+            attendances = calendar._attendance_intervals_batch(start_dt, end_dt, resources=self)
+            dict_attendance_per_calendar[calendar] = attendances
+
+        for res in self:
+            leave = personal_leaves_intervals | global_leave_per_calendar[res.calendar_id]
+            resource_work_intervals = res._leaves_interval_gantt_view(start_datetime, end_datetime, leave, dict_attendance_per_calendar[res.calendar_id][res.id])
+            work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
+            # start + flatten(intervals) + end
+            work_intervals = [start_datetime] + list(chain.from_iterable(work_intervals)) + [end_datetime]
+            # put it back to UTC
+            work_intervals = list(map(lambda dt: dt.astimezone(utc), work_intervals))
+            # pick groups of two
+            work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
+            leaves_mapping[res.id] = work_intervals
+        return leaves_mapping
+
+    def _personal_leaves_intervals(self, start_dt, end_dt):
+        resource_ids = [r.id for r in self]
+        domain = [
+            ('time_type', '=', 'leave'),
+            ('resource_id', 'in', resource_ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        return self.env['resource.calendar.leaves'].search(domain)
```

Comment:
```suggestion
        return self.env['resource.calendar.leaves'].search([
            ('time_type', '=', 'leave'),
            ('resource_id', 'in', self.ids),
            ('date_from', '<=', end_dt),
            ('date_to', '>=', start_dt),
        ])
```
1) Create variable if you use the same value multiple times or as a way to fragment/document your code when it is getting complex. In this case, it really isn't worth it.
2) `[r.id for r in self]` = `self.ids` :scream: